Code/Framework/AzFramework/AzFramework/Network/IRemoteTools.h

```diff
@@ -0,0 +1,252 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#pragma once
+
+#include <AzCore/base.h>
+#include <AzCore/EBus/Event.h>
+#include <AzCore/Memory/OSAllocator.h>
+#include <AzCore/Memory/SystemAllocator.h>
+#include <AzCore/RTTI/ReflectContext.h>
+#include <AzCore/RTTI/RTTI.h>
+#include <AzCore/Serialization/SerializeContext.h>
+#include <AzCore/std/containers/deque.h>
+#include <AzCore/std/containers/unordered_map.h>
+#include <AzCore/std/smart_ptr/intrusive_ptr.h>
+#include <AzCore/std/smart_ptr/intrusive_refcount.h>
+#include <AzCore/std/string/string.h>
+
+namespace AZ
+{
+    template<typename T>
+    class Interface;
+
+    class Name;
+}
+
+namespace AzFramework
+{
+    class RemoteToolsMessage : public AZStd::intrusive_refcount<size_t>
+    {
+    public:
+        AZ_CLASS_ALLOCATOR(RemoteToolsMessage, AZ::OSAllocator, 0);
+        AZ_RTTI(RemoteToolsMessage, "{8512328C-949D-4F0C-B48D-77C26C207443}");
+
+        RemoteToolsMessage()
+            : m_msgId(0)
+            , m_senderTargetId(0)
+            , m_customBlob(nullptr)
+            , m_customBlobSize(0)
+            , m_isBlobOwner(false)
+            , m_immediateSelfDispatch(false)
+        {
+        }
+        RemoteToolsMessage(AZ::u64 msgId)
+            : m_msgId(msgId)
+            , m_senderTargetId(0)
+            , m_customBlob(nullptr)
+            , m_customBlobSize(0)
+            , m_isBlobOwner(false)
+            , m_immediateSelfDispatch(false)
+        {
+        }
+
+        virtual ~RemoteToolsMessage();
+
+        //! Add a custom data blob to a Remote Tools message
+        //! @param blob Pointer to the start of the data blob
+        //! @param blobSize Size of the data blob
+        //! @param ownBlob Whether this message owns the life cycle of the blob memory
+        void AddCustomBlob(const void* blob, size_t blobSize, bool ownBlob = false);
+
+        //! Sets if this message can be immediately dispatched if the receipient is itself
+        //! @param immediateSelfDispatchEnabled Whether this message can be immediately dispatched
+        void SetImmediateSelfDispatchEnabled(bool immediateSelfDispatchEnabled);
+
+        //! Gets if this message can be immediately dispatched if the receipient is itself
+        //! @return Whether this message can be immediately dispatched
+        bool IsImmediateSelfDispatchEnabled() const;
+
+        //! Gets the custom data blob assosciated with this message
+        //! @return Memory address of the start of the blob or nullptr if there is none
+        const void* GetCustomBlob() const;
+
+        //! Gets the size of the custom data blob associated with this message
+        //! @return Size of the data blob or 0 if there is none
+        size_t GetCustomBlobSize() const;
```

Comment:
Instead of having two separate functiosn for getting the custom blob pointer and size.
You could just have one function that returns an `AZStd::span<AZStd::byte>`.

I think that better encapsulates the intent the custom blob is a container of bytes
```suggestion
        //! Gets the custom data blob assosciated with this message
        //! @return span of the custom blob
        AZStd::span<AZStd::byte const> GetCustomBlob() const;
```

Reply:
I like the idea of representing it with a span. I'm going to keep the Size getter for its convenience as a logical gate.

Reply:
I've updated the internals to use a span and added the span getter. Let me know if any of that usage seems inadvisable.

Reply:
The span changes look good.